src/sys/socket/addr.rs

```diff
@@ -1761,6 +1761,89 @@ pub mod netlink {
     }
 }
 
+#[cfg(target_os = "linux")]
+#[cfg_attr(docsrs, doc(cfg(all())))]
+pub mod can {
+    use super::*;
+    use crate::sys::socket::addr::AddressFamily;
+    use libc::{sa_family_t, sockaddr_can, c_int, c_uint};
+    use std::{fmt, mem};
+
+    /// Address for the Linux SocketCAN implementation for the Controller Area Network
+    /// automotive bus protocol.
+    ///
+    /// # References
+    ///
+    /// [SocketCAN](https://docs.kernel.org/networking/can.html)
+    #[derive(Copy, Clone)]
+    #[allow(missing_debug_implementations)]
+    #[repr(transparent)]
+    pub struct CanAddr(pub(in super::super) sockaddr_can);
+
+    impl CanAddr {
+        /// Construct a new CAN socket address from its interface index.
+        pub fn new(ifindex: c_uint) -> Self {
+            let mut addr: sockaddr_can = unsafe { mem::zeroed() };
+            addr.can_family = AddressFamily::Can as sa_family_t;
+            addr.can_ifindex = ifindex as c_int;
+
+            Self(addr)
+        }
+
+        /// Return the socket's CAN interface index.
+        pub const fn ifindex(&self) -> c_uint {
+            self.0.can_ifindex as c_uint
+        }
+    }
+
+    impl private::SockaddrLikePriv for CanAddr {}
+    impl SockaddrLike for CanAddr {
+        unsafe fn from_raw(
+            addr: *const libc::sockaddr,
+            len: Option<libc::socklen_t>,
+        ) -> Option<Self>
+        where
+            Self: Sized,
+        {
+            if let Some(l) = len {
+                if l != mem::size_of::<libc::sockaddr_can>() as libc::socklen_t {
+                    return None;
+                }
+            }
+            if (*addr).sa_family as i32 != libc::AF_CAN {
+                return None;
+            }
+            Some(Self(ptr::read_unaligned(addr as *const _)))
+        }
+    }
+
+    impl AsRef<libc::sockaddr_can> for CanAddr {
+        fn as_ref(&self) -> &libc::sockaddr_can {
+            &self.0
+        }
+    }
+
+    impl fmt::Display for CanAddr {
+        fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+            write!(f, "ifindex: {}", self.ifindex())
+        }
+    }
+
+    impl PartialEq for CanAddr {
+        fn eq(&self, other: &CanAddr) -> bool {
+            self.ifindex() == other.ifindex()
+        }
+    }
+
+    impl Eq for CanAddr {}
+
+    impl Hash for CanAddr {
```

Comment:
As stated above, this implementation works for the common/base CAN address in which the union is zeroed out and ignored; and the only thing that could change from one address to another is the value of the index.

But if we wanted to implement the two other protocols in this struct (J1939 and ISO-TP), then, since the struct  is Sized, could we just do a bitwise comparison for  `PartialEq`? And a hash of the bytes as a slice?

Would that work?

Reply:
What if you get a sockaddr from a function like `getsockname`?  Then it might match any of these union variants, or even none of them.  Implementing `Hash` by casting to a byte slice would work only if you can guarantee that all of the memory will always be initialized.  That should be the case for constructors like `getsockname`, but what about manual construction?

Reply:
Yes, again, I was assuming this is not a general solution for all enums, but acceptable for this type of `sockaddr` where the convention is to zero out the struct.

But agree this is a cheat. The alternate implementation of separate implementations for each enum possibility would fix this.